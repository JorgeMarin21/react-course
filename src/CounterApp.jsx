import PropTypes from 'prop-types';
import { useState } from 'react';

export const CounterApp = ( { value } ) => {
  if ( !value && value !== 0 )
  throw new Error( 'No existe valor');

  const [ counter, setCounter ] = useState(value);

  const handleAdd = () => {
    setCounter(counter + 1 )
    // setCounter( (c) => c + 1 )
  }
  const handleSubstractAdd = () => {
    setCounter(counter - 1 )
    // setCounter( (c) => c - 1 )
  }
  const handleReset = () => {
    setCounter( value )
    // setCounter( (c) => c - 1 )
  }

  return (
    <>
      <h1> CounterApp </h1>
      <h2> { counter } </h2>
      <button aria-label="btn-plus" onClick={ handleAdd }>+1</button>
      <button aria-label="btn-minus" onClick={ handleSubstractAdd }>-1</button>
      <button aria-label="btn-reset" onClick={ handleReset }>Reset</button>
    </>
  )
}

CounterApp.propTypes = {
  value: PropTypes.number.isRequired
}