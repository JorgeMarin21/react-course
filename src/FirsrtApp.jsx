import PropTypes from 'prop-types'
const getMessage = () => 'Hola Mijo'

export const FirstApp = ( { title, subtitle = 'vamos' } ) => {

  if ( !title )
  throw new Error( 'prueba')
  return (
    <>
      <h1 data-testid='test-title'>{title}</h1>
      <p> { subtitle } </p>
    </>
  )
}

FirstApp.propTypes = {
  title: PropTypes.string.isRequired,
  subtitle: PropTypes.string
}

// FirstApp.defaultProps = {
//   title: 'No hay titulo'
// }
// defaultProps Deprecated