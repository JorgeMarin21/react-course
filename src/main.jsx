import React from 'react';
import ReactDOM from 'react-dom/client';
// import { HelloWorldApp } from './HelloWorldApp';
// import { CounterApp } from './CounterApp';
import './styles.css';
// ReactDOM.createRoot(document.getElementById('root')).render(
//     <React.StrictMode>
//         <FirstApp title = 'El titulo' subtitle = 'El subtitulo' />
//     </React.StrictMode>
// );
ReactDOM.createRoot(document.getElementById('root')).render(
    <React.StrictMode>
        <CounterApp value = { 0 } />
    </React.StrictMode>
);