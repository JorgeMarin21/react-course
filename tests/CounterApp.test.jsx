import { fireEvent, render, screen } from "@testing-library/react";
import { CounterApp } from "../src/CounterApp";

describe('Counter App', () => { 
    const initialValue = 0;
    test('should match with the Snapshot', () => { 
        const { container } = render(<CounterApp value={initialValue}/>);
        expect(container).toMatchSnapshot();
     });

     test('should show 100 as initial value', () => { 
         render(<CounterApp value = {100} />);
         expect( screen.getByText(100) ).toBeTruthy();
         expect( screen.getByRole('heading', {level: 2}).innerHTML ).toContain('100'); // more accurate
     });

     test('should increment with +1 button', () => {
        render(<CounterApp value = {initialValue} />);
        fireEvent.click( screen.getByText('+1') );
        expect(screen.getByText('1')).toBeTruthy();
        expect( screen.getByRole('heading', {level: 2}).innerHTML ).toContain('1'); // more accurate
     });

     test('should decrement with +1 button', () => {
        render(<CounterApp value = {initialValue} />);
        fireEvent.click( screen.getByText('-1') );
        expect(screen.getAllByText('-1').length).toBe(2);
         expect( screen.getByRole('heading', {level: 2}).innerHTML ).toContain('-1'); // more accurate
     });

     test('should reset the counter', () => {
        render(<CounterApp value = {initialValue} />);
        fireEvent.click( screen.getByText('+1') );
        fireEvent.click( screen.getByText('+1') );
        fireEvent.click( screen.getByText('+1') );
        expect( screen.getByRole('heading', {level: 2}).innerHTML ).toContain('3'); 
        fireEvent.click( screen.getByText('Reset') );
        expect(screen.getByText('0')).toBeTruthy();
        expect( screen.getByRole('heading', {level: 2}).innerHTML ).toContain('0');
     });

     test('should reset the counter - second way', () => {
        render(<CounterApp value = {initialValue} />);
        fireEvent.click( screen.getByRole('button', {name: 'btn-plus'}) );
        fireEvent.click( screen.getByRole('button', {name: 'btn-plus'}) );
        expect( screen.getByRole('heading', {level: 2}).innerHTML ).toContain('2'); 
        fireEvent.click( screen.getByRole('button', {name: 'btn-reset'}) );
        expect(screen.getByText('0')).toBeTruthy();
        expect( screen.getByRole('heading', {level: 2}).innerHTML ).toContain('0');
     });
 });