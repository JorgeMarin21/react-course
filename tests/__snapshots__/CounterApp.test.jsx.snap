// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter App should match with the Snapshot 1`] = `
<div>
  <h1>
     CounterApp 
  </h1>
  <h2>
     
    0
     
  </h2>
  <button
    aria-label="btn-plus"
  >
    +1
  </button>
  <button
    aria-label="btn-minus"
  >
    -1
  </button>
  <button
    aria-label="btn-reset"
  >
    Reset
  </button>
</div>
`;
